### DISCUSSION TITLE
Generate interface for model

### DISCUSSION THREAD

**Discussion by hikyaru-suzuki** [aarondl/sqlboiler#886] [Ideas]
Now, we are [taking a position](https://github.com/DATA-DOG/go-sqlmock) to use [Datadog's sql mock library](https://github.com/DATA-DOG/go-sqlmock) for model mocking.

it's a common usecase useing sqlboiler.
```go
users, err := models.Users().All(ctx, exec)
```
In this case, `users` variable is array of `User` struct pointer and programmer don't know details of executed query. This is good point of ORM.
But, sqlmock need to know raw query for mocking. I don't want to touch query and only would like to touch abstracted API.

So, I propose to add interface for model method and then we can generate that's mock code for testing by third party mock generation library. We will be able to describe test code without knowing raw query!
Please your opinion!

**Comment by aarondl** [comment]
sqlmock afaik doesn't require you to know the query. You can say `ExpectExec(".*")` and simply use that to return your user object results.

I'd recommend you do not mock your data layer though. It's generally going to be better to just test directly against the database. It's just such a core part of your app, unless you have other integration tests that do it you'd be missing a lot of proper execution of the code.
